File: examples/android/MSDKSampleAppJava/lib/src/main/java/msdk/java/types/PoolConfig.java
package msdk.java.types;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.RawRes;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;

import msdk.java.logger.Logger;
import msdk.java.utils.CommonUtils;

public class PoolConfig {
    public static PoolConfigBuilder builder() {
        return new PoolConfigBuilder();
    }

    public static File writeGenesisFile(Context context, Integer genesisPoolResId) {
        File genesisFile = new File(CommonUtils.getRootDir(context), "pool_transactions_genesis");
        if (!genesisFile.exists()) {
            try (FileOutputStream stream = new FileOutputStream(genesisFile)) {
                Logger.getInstance().d("writing poolTxnGenesis to file: " + genesisFile.getAbsolutePath());
                if (genesisPoolResId != null) {
                    try (InputStream genesisStream = context.getResources().openRawResource(genesisPoolResId)) {
                        byte[] buffer = new byte[8 * 1024];
                        int bytesRead;
                        while ((bytesRead = genesisStream.read(buffer)) != -1) {
                            stream.write(buffer, 0, bytesRead);
                        }
                    }
                }
            } catch (Exception ex) {
                ex.printStackTrace();
            }
        }
        return genesisFile;
    }

    public static final class PoolConfigBuilder {
        private String genesisPath;
        private String poolName;

        private PoolConfigBuilder() {
        }

        /**
         * Set genesis transactions path
         *
         * @param genesisPath path to genesis transactions
         * @return this
         */
        public @NonNull
        PoolConfigBuilder withgGenesisPath(@NonNull String genesisPath) {
            this.genesisPath = genesisPath;
            return this;
        }

        /**
         * Set pool name
         *
         * @param poolName pool name
         * @return this
         */
        public @NonNull
        PoolConfigBuilder withgPoolName(@NonNull String poolName) {
            this.poolName = poolName;
            return this;
        }

        /**
         * Creates pool initialization config
         *
         * @return pool initialization config
         */
        public @NonNull
        String build() throws JSONException {
            JSONObject poolConfig = new JSONObject();
            poolConfig.put("genesis_path", this.genesisPath);
            poolConfig.put("pool_name", this.poolName);
            return poolConfig.toString();
        }
    }
}

File: examples/android/MSDKSampleAppKotlin/app/src/main/java/msdk/kotlin/sample/types/PoolConfig.kt
package msdk.kotlin.sample.types

import android.content.Context
import msdk.kotlin.sample.logger.Logger.Companion.instance
import msdk.kotlin.sample.utils.CommonUtils
import org.json.JSONException
import org.json.JSONObject
import java.io.File
import java.io.FileOutputStream

object PoolConfig {
    fun builder(): PoolConfigBuilder {
        return PoolConfigBuilder()
    }

    fun writeGenesisFile(context: Context, genesisPoolResId: Int?): File {
        val genesisFile =
            File(CommonUtils.getRootDir(context), "pool_transactions_genesis")
        if (!genesisFile.exists()) {
            try {
                FileOutputStream(genesisFile).use { stream ->
                    instance.d("writing poolTxnGenesis to file: " + genesisFile.absolutePath)
                    if (genesisPoolResId != null) {
                        context.resources.openRawResource(genesisPoolResId)
                            .use { genesisStream ->
                                val buffer = ByteArray(8 * 1024)
                                var bytesRead: Int
                                while (genesisStream.read(buffer)
                                        .also { bytesRead = it } != -1
                                ) {
                                    stream.write(buffer, 0, bytesRead)
                                }
                            }
                    }
                }
            } catch (ex: Exception) {
                ex.printStackTrace()
            }
        }
        return genesisFile
    }

    class PoolConfigBuilder {
        private var genesisPath: String? = null
        private var poolName: String? = null

        /**
         * Set genesis transactions path
         *
         * @param genesisPath path to genesis transactions
         * @return this
         */
        fun withgGenesisPath(genesisPath: String): PoolConfigBuilder {
            this.genesisPath = genesisPath
            return this
        }

        /**
         * Set pool name
         *
         * @param poolName pool name
         * @return this
         */
        fun withgPoolName(poolName: String): PoolConfigBuilder {
            this.poolName = poolName
            return this
        }

        /**
         * Creates pool initialization config
         *
         * @return pool initialization config
         */
        @Throws(JSONException::class)
        fun build(): String {
            val poolConfig = JSONObject()
            poolConfig.put("genesis_path", genesisPath)
            poolConfig.put("pool_name", poolName)
            return poolConfig.toString()
        }
    }
}